File: csharp/perfect-numbers/Iteration1/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number <= 0) throw new ArgumentOutOfRangeException();
        
        var aliquot = number
            .Factors()
            .Where(n => n != number)
            .Sum();

        if (number == aliquot) return Classification.Perfect;
        if (number < aliquot) return Classification.Abundant;
        return Classification.Deficient;
    }

    public static IEnumerable<int> Factors(this int number) =>
        Enumerable
            .Range(1, (int)Math.Sqrt(number))
            .Where(n => number % n == 0)
            .Select(n => new[] { n, (number / n) })
            .SelectMany(n => n)
            .Distinct();
}

File: zig/perfect-numbers/Iteration1/perfect_numbers.zig

pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

fn aliquotSum(n: u64) u64 {
    var factor = n - 1;
    var sum: u64 = 0;

    while (factor > 0) {
        if (n % factor == 0) {
            sum += factor;
        }

        factor -= 1;
    }

    return sum;
}

pub fn classify(n: u64) Classification {
    const sum = aliquotSum(n);

    if      (sum == n) { return Classification.perfect; }
    else if (sum > n)  { return Classification.abundant; }
    else               { return Classification.deficient; }
}
